subprojects/core/src/integTest/groovy/org/gradle/api/provider/ProviderIntegrationTest.groovy

```diff
@@ -0,0 +1,381 @@
+            }
+        """
+        succeeds('myTask')
+
+        then:
+        !defaultOutputFile.exists()
+        customOutputFile.isFile()
+        customOutputFile.text == OUTPUT_FILE_CONTENT
+
+        where:
+        language | taskTypeImpl
+        'groovy' | customGroovyBasedTaskType()
+        'java'   | customJavaBasedTaskType()
+    }
+
+    def "can assign default values for provider properties"() {
+        given:
+        buildFile << """
+            task myTask(type: MyTask)
+
+            class MyTask extends DefaultTask {
+                private Provider<Boolean> enabled = project.defaultProvider(Boolean)
+                private Provider<FileCollection> outputFiles = project.defaultProvider(FileCollection)
+                
+                @Input
+                boolean getEnabled() {
+                    enabled.get()
+                }
+                
+                void setEnabled(Provider<Boolean> enabled) {
+                    this.enabled = enabled
+                }
+                
+                @OutputFiles
+                FileCollection getOutputFiles() {
+                    outputFiles.get()
+                }
+
+                void setOutputFiles(Provider<FileCollection> outputFiles) {
+                    this.outputFiles = outputFiles
+                }
+                
+                @TaskAction
+                void resolveValue() {
+                    if (getEnabled()) {
+                        getOutputFiles().each {
+                            it.text = '$OUTPUT_FILE_CONTENT'
+                        }
+                    }
+                }
+            }
+        """
+
+        when:
+        succeeds('myTask')
+
+        then:
+        !defaultOutputFile.exists()
+    }
+
+    @Unroll
+    def "can lazily map extension property value to task property with #mechanism"() {
+        given:
+        buildFile << pluginWithExtensionMapping { taskConfiguration }
+        buildFile << customGroovyBasedTaskType()
+
+        when:
+        succeeds('myTask')
+
+        then:
+        !defaultOutputFile.exists()
+        customOutputFile.isFile()
+        customOutputFile.text == OUTPUT_FILE_CONTENT
+
+        where:
+        mechanism            | taskConfiguration
+        'convention mapping' | taskConfiguredWithConventionMapping()
+        'provider'           | taskConfiguredWithProvider()
+    }
+
+    def "can use provider to define a task dependency"() {
+        given:
+        buildFile << """
+            task producer {
+                ext.destFile = file('test.txt')
+                outputs.file destFile
+
+                doLast {
+                    destFile << '$OUTPUT_FILE_CONTENT'
+                }
+            }
+
+            def targetFile = provider { producer.destFile }
+            
+            targetFile.builtBy producer
+
+            task consumer {
+                dependsOn targetFile
```

Comment:
Doesn't this fail during configuration time due to file validation? I think it would try to validate the existence of the file but it does not exist yet as the file is created during execution time in the action of `producer`.